job_definitions/release_pipeline.yml

```diff
@@ -59,6 +61,81 @@
           }
         }
 
+        def run_visual_regression_tests(stage) {
+          def job_name = "visual-regression-${stage}"
+          def vr_job = null
+
+          waitUntil {
+            node {
+              vr_job = build job: job_name, parameters: [
+                string(name: "COMMAND", value: "test"),
+              ], propagate: false
+            }
+
+            def vr_job_result = vr_job.result
+            def vr_job_number = vr_job.number
+            vr_job = null
+
+            if (vr_job_result == 'SUCCESS') {
+                return true
+            } else {
+              echo "Visual regression tests build result: ${vr_job_result}"
+
+              timeout(time:1, unit:'DAYS') {
+                script {
+                  env.FAILED_TEST_ACTION = input message: 'Visual regression tests failed. Retry, approve, bypass?', parameters: [
+                    [
+                      $class: 'ChoiceParameterDefinition',
+                      choices: 'Retry\nApprove\nBypass',
+                      description: 'Failed visual regression tests action',
+                      name: 'action'
+                    ]
+                  ]
+                }
+
+                echo "Action chosen was ${env.FAILED_TEST_ACTION}"
+              }
+
+              if (env.FAILED_TEST_ACTION == 'Retry') {
+                echo "Retrying visual regression tests on ${stage}."
+                return false
+
+              } else if (env.FAILED_TEST_ACTION == 'Approve') {
+                def test_job = Jenkins.instance.getItemByFullName(job_name)
+                def test_job_last_build_number = test_job.getLastBuild().getNumber()
+                test_job = null
+
+                echo "Build triggered by this pipeline run: ${vr_job_number}. Last build: ${test_job_last_build_number}"
+
+                if (test_job_last_build_number == vr_job_number) {
+                  echo "Verified that the last visual regression job built was triggered by this pipeline. Approving."
+                  try {
+                    node {
+                      build job: job_name, parameters: [
+                        string(name: "COMMAND", value: "approve"),
+                      ]
+                    }
+
+                  } catch(inside_error) {
+                    echo "Approving visual regression tests failed: ${inside_error}"
+                    return false
+                  }
+
+                  return true
+                } else {
+                  input(message: "Visual regression tests have run elsewhere since this pipeline triggered the job. They will be re-run here when you proceed.")
+                  return false
+                }
+              } else if (env.FAILED_TEST_ACTION == 'Bypass') {
```

Comment:
When would this step every be useful?

Reply:
I don't have a specific use case in mind, but we have the same option on the functional tests, so presumably there's a use case somewhere.

Reply:
Yeah, I guess having the flexibility is a good thing!

Reply:
One possible example: someone's testing a framework in a different state (e.g. g-cloud 8 open), and the only changes failing a report are from the banners that appear on the homepage/dashboards. In this case, the framework might be open for an indeterminate amount of time, so you can't really wait to re-run them, but you can't approve the tests either. Bypassing them at that point might be OK as long as the developer was sure no unintended changes slipped through.